mlflow/models/utils.py

```diff
@@ -821,6 +821,17 @@ def _enforce_mlflow_datatype(name, values: pd.Series, t: DataType):
     if is_upcast:
         return values.astype(numpy_type, errors="raise")
     else:
+        # support converting long -> float/double for 0 values
+        def all_zero_or_ones(xs):
+            return all(pd.isnull(x) or x in [0, 1] for x in xs)
```

Comment:
if we want to do something like all int32s
```suggestion
        # support converting int32 -> float/double for 0 values
        def all_zero_or_ones(xs):
            return all(pd.isnull(x) or x < 2 ** 31 for x in xs)
```

Reply:
but i guess it's probably safer to just keep 0 or 1

Reply:
shouldn't be `abs(x) < 2 ** 31` ?

Reply:
another issue is it uses for loop which is very slow,( it causes big overhead in enforcing data type), and it has many other similar for loop in  enforcing dat code, can we improve it ? (can be in follow-up pr)

like using numpy batch operation:
`(np.abs(xs) < 2**31).all()`

Reply:
with vectorization:
```
(np.abs(xs) < 2**31).all()
```
without vectorization:
```
all(x < 2 ** 31 for x in xs)  # it iterates elements one by one
```

See https://www.programiz.com/python-programming/numpy/vectorization for more details of why vectorization making it faster.

Reply:
can be a follow-up work though, we have many places to do the similar change